### PR TITLE
[mlir][TilingInterface] Move TilingInterface tests to use transform dialect ops.

### PR DESCRIPTION
**Pull request by MaheshRavishankar**
In the process a couple of test transform dialect ops are added just
for testing. These operations are not intended to use as full flushed
out of transformation ops, but are rather operations added for testing.

A separate operation is added to `LinalgTransformOps.td` to convert a
`TilingInterface` operation to loops using the
`generateScalarImplementation` method implemented by the
operation. Eventually this and other operations related to tiling
using the `TilingInterface` need to move to a better place (i.e. out
of `Linalg` dialect)

### CODE REVIEW

**Comment by MaheshRavishankar** [comment]
This is based on top of https://github.com/llvm/llvm-project/pull/76871 . Only the last commit is part of this PR

File: mlir/test/lib/Interfaces/TilingInterface/TestTilingInterfaceTransformOps.cpp [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+#include "mlir/IR/Dominance.h"
+#include "mlir/IR/OpImplementation.h"
+#include "mlir/Interfaces/TilingInterface.h"
+
+#define GET_OP_CLASSES
+#include "TestTilingInterfaceTransformOps.h.inc"
+
+using namespace mlir;
+using namespace mlir::transform;
+
```
Comment (qedawkins) [inline mlir/test/lib/Interfaces/TilingInterface/TestTilingInterfaceTransformOps.cpp:30]: The testing ops seem like they could be generally useful; I would import the TODO from before as
```
// TODO: this file should disappear and the associated transform ops should be
// promoted for general use.
```

File: mlir/test/Interfaces/TilingInterface/tile-pad-using-interface.mlir [hunk 2]
```diff
@@ -137,4 +188,14 @@ func.func @static_pad_tensor_outer_tiling(%input_tensor: tensor<7x9xf32>,
     } {__internal_transform__ = "pad_inner_tiling"} : tensor<7x9xf32> to tensor<15x16xf32>
```
Comment (hanhanW) [inline mlir/test/Interfaces/TilingInterface/tile-pad-using-interface.mlir:188]: We don't need these attributes after switching to transform dialects. The new tests match the op directly.

File: mlir/test/lib/Interfaces/TilingInterface/lit.local.cfg [hunk 3]
```diff
@@ -0,0 +1 @@
+config.suffixes.remove(".td")
```
Comment (hanhanW) [inline mlir/test/lib/Interfaces/TilingInterface/lit.local.cfg:1]: do we need the file?

Comment (MaheshRavishankar) [reply mlir/test/lib/Interfaces/TilingInterface/lit.local.cfg:1]: Yeah, without this the lit test framework expects the .td file added here to contain tests (not sure why). This avoids it.

File: mlir/test/lib/Interfaces/TilingInterface/TestTilingInterfaceTransformOps.cpp [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+        continue;
+      worklist.push_back(producer);
+      producers.insert(producer);
+    }
+  }
+  return producers;
+}
+
+/// Apply a tiling transformation to all payload ops and store both the
+/// tiled operation as well as the created tile loops.
```
Comment (qedawkins) [inline mlir/test/lib/Interfaces/TilingInterface/TestTilingInterfaceTransformOps.cpp:55]: Comment looks copy pasted from below? Mention that it also fuses greedily and yields the fused ops.

File: mlir/test/lib/Interfaces/TilingInterface/TestTilingInterfaceTransformOps.cpp [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+
+//===----------------------------------------------------------------------===//
+// TestTileUsingForallOp
+//===----------------------------------------------------------------------===//
+
+/// Apply a tiling transformation to all payload ops and store both the
+/// tiled operation as well as the created tile loops.
+template <typename Range>
+static LogicalResult
+applyTileToAll(RewriterBase &rewriter, Operation *transformOp,
```
Comment (qedawkins) [inline mlir/test/lib/Interfaces/TilingInterface/TestTilingInterfaceTransformOps.cpp:169]: nit: `applyTileUsingForallToAll`?

File: mlir/test/lib/Interfaces/TilingInterface/TestTilingInterfaceTransformOps.td [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+include "mlir/IR/OpBase.td"
+
+def TestFuseAndYieldOp : Op<Transform_Dialect, "test.fuse_and_yield",
+  [FunctionalStyleTransformOpTrait, MemoryEffectsOpInterface,
+   DeclareOpInterfaceMethods<TransformOpInterface>,
+   ReportTrackingListenerFailuresOpTrait]> {
+  let description = [{
+    Tiles the operations pointed to by the target handle, fuses their
+    producers greedily using the options provided as attributes.
+    It also yields some of the fused producers for testing.
```
Comment (qedawkins) [inline mlir/test/lib/Interfaces/TilingInterface/TestTilingInterfaceTransformOps.td:26]: Add a description of the return modes, i.e.
```
On success returns the tiled operations as well as generated loops.
Emits a definite failure if tiling fails.
```

File: mlir/test/lib/Interfaces/TilingInterface/TestTilingInterfaceTransformOps.td [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+       [DeclareOpInterfaceMethods<TransformOpInterface>,
+        DeclareOpInterfaceMethods<MemoryEffectsOpInterface>,
+        ReportTrackingListenerFailuresOpTrait]> {
+  let description = [{
+    Test operation use to test tiling using TilingInterface and scf.forall for
+    the loop constructs. This is similar to
+    `transform.structured.tile_using_for`. Use of this operation is an
+    intermediate state and will be replaced in due course with either
+    `transform.structured.tile_using_for` or
+    `transform.structured.tile_using_forall`.
```
Comment (qedawkins) [inline mlir/test/lib/Interfaces/TilingInterface/TestTilingInterfaceTransformOps.td:52]: ditto, add a brief description of the return modes.

File: mlir/test/lib/Interfaces/TilingInterface/TestTilingInterfaceTransformOps.td [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+include "mlir/Dialect/SCF/IR/DeviceMappingInterface.td"
+include "mlir/Dialect/Transform/IR/TransformDialect.td"
+include "mlir/Dialect/Transform/IR/TransformInterfaces.td"
+include "mlir/Dialect/Transform/IR/TransformTypes.td"
+include "mlir/Interfaces/SideEffectInterfaces.td"
+include "mlir/IR/OpBase.td"
+
+def TestFuseAndYieldOp : Op<Transform_Dialect, "test.fuse_and_yield",
+  [FunctionalStyleTransformOpTrait, MemoryEffectsOpInterface,
```
Comment (qedawkins) [inline mlir/test/lib/Interfaces/TilingInterface/TestTilingInterfaceTransformOps.td:20]: Why not declare the `MemoryEffects` interface methods here + `getEffects`?

**Review by qedawkins — APPROVED** [review]
Mostly looks good to me, getting rid of test passes is +1. It would be nice to phase out test ops over time too but it makes sense to me to go step by step and not introduce something unstable.

File: mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td [hunk 9]
```diff
@@ -293,7 +293,10 @@ def FuseOp : Op<Transform_Dialect, "structured.fuse",
   let results = (outs TransformHandleTypeInterface:$transformed,
                       Variadic<TransformHandleTypeInterface>:$loops);
 
-  let hasCustomAssemblyFormat = 1;
+  let assemblyFormat = [{
+    $target ($tile_sizes^)? (`interchange` $tile_interchange^)?
+    attr-dict `:` functional-type(operands, results)
+  }];
```
[the remark this answers is not in the document]
Comment (MaheshRavishankar) [reply mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td:299]: Its an NFC. None of the roundtrip tests failed.

File: mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp [hunk 10]
```diff
@@ -283,10 +283,7 @@ mlir::scf::tileUsingSCFForOp(RewriterBase &rewriter, TilingInterface op,
   // 1. Get the range of the loops that are represented by the operation.
   SmallVector<Range> iterationDomain = op.getIterationDomain(rewriter);
   size_t numLoops = iterationDomain.size();
-  if (numLoops == 0) {
-    return rewriter.notifyMatchFailure(
-        op, "unable to tile op with no iteration domain");
-  }
+
```
[the remark this answers is not in the document]
Comment (MaheshRavishankar) [reply mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:286]: Yeah, I am doing that in a follow up. I have a change that I have locally that reworks the logic here which I will send out after this lands. For now this just avoids an error.

File: mlir/test/lib/Interfaces/TilingInterface/TestTilingInterfaceTransformOps.cpp (hunk 1 above)
Comment (MaheshRavishankar) [reply mlir/test/lib/Interfaces/TilingInterface/TestTilingInterfaceTransformOps.cpp:30]: Added this TODO to the .td file.

File: mlir/test/lib/Interfaces/TilingInterface/TestTilingInterfaceTransformOps.cpp (hunk 5 above)
Comment (MaheshRavishankar) [reply mlir/test/lib/Interfaces/TilingInterface/TestTilingInterfaceTransformOps.cpp:169]: Seems a  bit wordy... Ill leave it as is for now.

File: mlir/test/lib/Interfaces/TilingInterface/TestTilingInterfaceTransformOps.td (hunk 8 above)
Comment (MaheshRavishankar) [reply mlir/test/lib/Interfaces/TilingInterface/TestTilingInterfaceTransformOps.td:25]: I just copy pasted and modified the `structured.fused` operation.

**Pull request merged**